Remotion/Web/ClientScript/Scripts/SmartPage.ts

```diff
@@ -0,0 +1,1307 @@
+    this._smartFocusFieldID = smartFocusFieldID;
+
+    this.AttachPageLevelEventHandlers();
+  }
+
+  public set_AbortQueuedSubmit (value: boolean): void
+  {
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('value', value);
+    this._abortQueuedSubmit = value;
+  };
+
+  public set_EventHandlers (eventHandlers: string[][]): void
+  {
+    ArgumentUtility.CheckTypeIsObject('eventHandlers', eventHandlers);
+    this._eventHandlers = eventHandlers;
+  };
+
+  public set_TrackedIDs (trackedIDs: string[]): void
+  {
+    ArgumentUtility.CheckTypeIsObject('trackedIDs', trackedIDs);
+    this._trackedIDs = trackedIDs;
+  };
+
+  public set_SynchronousPostBackCommands (synchronousPostBackCommands: string[]): void
+  {
+    ArgumentUtility.CheckTypeIsObject('synchronousPostBackCommands', synchronousPostBackCommands);
+    this._synchronousPostBackCommands = synchronousPostBackCommands;
+  };
+
+  // Attaches the event handlers to the page's events.
+  private AttachPageLevelEventHandlers()
+  {
+    this.RemoveEventHandler(window, 'load', this._loadHandler);
+    this.AddEventHandler(window, 'load', this._loadHandler);
+
+    // IE, Mozilla 1.7, Firefox 0.9
+    window.onbeforeunload = this._beforeUnloadHandler;
+
+    window.onunload = this._unloadHandler;
+
+    this.RemoveEventHandler(window, 'scroll', this._scrollHandler);
+    this.AddEventHandler(window, 'scroll', this._scrollHandler);
+
+    this.RemoveEventHandler(window.document, 'mousedown', this._mouseDownHandler);
+    this.AddEventHandler(window.document, 'mousedown', this._mouseDownHandler);
+
+    this.RemoveEventHandler(window.document, 'mouseup', this._mouseUpHandler);
+    this.AddEventHandler(window.document, 'mouseup', this._mouseUpHandler);
+
+    PageUtility.Instance.RegisterResizeHandler('#' + this._theForm.id, this._resizeHandler);
+
+    this._aspnetFormOnSubmit = this._theForm.onsubmit as Nullable<() => boolean>;
+    this._theForm.onsubmit = this._formSubmitHandler;
+    this._theForm.onclick = this._formClickHandler;
+    if (TypeUtility.IsDefined(window.__doPostBack))
+    {
+      this._aspnetDoPostBack = window.__doPostBack;
+      window.__doPostBack = this._doPostBackHandler;
+    }
+  };
+
+
+  // Called after page's html content is complete.
+  // Used to perform initalization code that only requires complete the HTML source but not necessarily all images.
+  public OnStartUp (isAsynchronous: boolean, isDirty: boolean): void
+  {
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('isAsynchronous', isAsynchronous);
+    ArgumentUtility.CheckNotNullAndTypeIsBoolean('isDirty', isDirty);
+
+    this._isDirty = isDirty;
+
+    if (this._isDirtyStateTrackingEnabled)
+      this.AttachDataChangedEventHandlers();
+
+    var pageRequestManager = this.GetPageRequestManager();
+    if (pageRequestManager != null)
+    {
+      (Sys.WebForms.PageRequestManager.prototype as any)._updatePanel = this.Sys$WebForms$PageRequestManager$_updatePanel;
```

Comment:
Sadly, there is no real way of improving the type safety here since it is not possible to override existing declarations (in this case the ASP.NET AJAX type declarations).

Reply:
@a-ctor there's also not an option to do a cast to a special type declaration that just contains the _updatePanel field?

Reply:
Well yes but this does not provide any more type safety than the current solution since you are just writing out the function signature a second time. But I can change it to such a cast if you want.

Reply:
The idea is that we put the external dependencies all in one place instead of having the knowledge about it sprinkled in the code. Right now, there's just one usage and it hopefully stays that way, but we could establish a precent and document this way how we deal with these kind of situations.